### PR TITLE
Switch to libjuju and asyncio

### LINKED ISSUE #742: Juju Bootstrap to MAAS should not do --debug
I certainly do like --debug output, but I think most users would not want to see it, especially when it contains errors like this (even though I know they are transient errors):

```
          Attempting to connect to 10.5.29.230:22
          19:49:45 DEBUG juju.provider.common bootstrap.go:392 connection attempt for 10.5.29.230 failed: ssh: connect to host 10.5.29.230 port
          19:49:51 DEBUG juju.provider.common bootstrap.go:392 connection attempt for 10.5.29.230 failed: Warning: Permanently added '10.5.29.23
          /var/lib/juju/nonce.txt does not exist
          19:49:56 DEBUG juju.provider.common bootstrap.go:392 connection attempt for 10.5.29.230 failed: /var/lib/juju/nonce.txt does not exist
          19:50:01 DEBUG juju.provider.common bootstrap.go:392 connection attempt for 10.5.29.230 failed: /var/lib/juju/nonce.txt does not exist
```

I suggest we do not bootstrap --debug by default, but only do so if a --debug flag is present on the conjure-up command line.

### PR DESCRIPTION
**Pull request by johnsca**
Large refactor to switch from macumba and threading to libjuju and asyncio.

Also resolves #742

### CODE REVIEW

**Comment by battlemidget** [comment]
My first run attempted to select an existing model via conjure-up and was greeted with this error:

```
2017-03-06 21:39:50,761 [INFO] conjure-up/ghost - juju.py:149 - Connecting to model conjure-up-localhost-8d1:conjure-up-ghost-465...
2017-03-06 21:39:50,768 [ERROR] conjure-up/ghost - __init__.py:16 - 'conjure-up-ghost-465'
Traceback (most recent call last):
  File "/home/ubuntu/conjure-up/conjureup/ui/__init__.py", line 13, in show_exception_message
    raise ex
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/ubuntu/conjure-up/conjureup/controllers/deploy/gui.py", line 60, in _pre_deploy_exec
    app.env['JUJU_PROVIDERTYPE'] = model_info().provider_type
  File "/home/ubuntu/conjure-up/conjureup/juju.py", line 32, in _decorator
    login(force=True)
  File "/home/ubuntu/conjure-up/conjureup/juju.py", line 179, in login
    f.result()  # block until connected
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 405, in result
    return self.__get_result()
  File "/usr/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/home/ubuntu/conjure-up/conjureup/juju.py", line 150, in _model_connect
    await app.juju.client.connect_model(model_name)
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/model.py", line 405, in connect_model
    self.connection = await connection.Connection.connect_model(model_name)
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/client/connection.py", line 270, in connect_model
    model_uuid = models['models'][model_name]['uuid']
KeyError: 'conjure-up-ghost-465'
2017-03-06 21:39:50,772 [DEBUG] conjure-up/ghost - __init__.py:33 - Showing dialog for exception: 'conjure-up-ghost-465'
2017-03-06 21:39:50,883 [INFO] conjure-up/ghost - juju.py:167 - Disconnecting model
2017-03-06 21:39:50,883 [INFO] conjure-up/ghost - juju.py:169 - Disconnected
2017-03-06 21:39:51,677 [INFO] conjure-up/ghost - eventloop: ev.py:40 - Stopping eventloop
2017-03-06 21:39:51,678 [INFO] conjure-up/ghost - app.py:371 - Shutting down
```

File: conjureup/app.py [hunk 1]
```diff
@@ -43,7 +44,7 @@ def parse_options(argv):
                         spell, or a keyword matching multiple spells
                         ('openstack')""")
     parser.add_argument('-d', '--debug', action='store_true',
-                        dest='debug', default=True,
+                        dest='debug', default=False,
```
Comment (battlemidget) [inline conjureup/app.py]: Attach the fixes bug for https://github.com/conjure-up/conjure-up/issues/742

File: conjureup/ui/views/deploystatus.py [hunk 2]
```diff
@@ -18,19 +15,31 @@ def __init__(self, app):
...
         """
-        EventLoop.loop.event_loop._loop.call_soon_threadsafe(
-            self._refresh_nodes_on_main_thread)
-
-    def _refresh_nodes_on_main_thread(self):
-        status = model_status()
-        for name, service in sorted(status['applications'].items()):
-            service_w = ServiceWidget(name, service)
+        for name, application in sorted(applications.items()):
+            # XXX refactor ubuntui to accept libjuju objects directly
```
Comment (battlemidget) [inline conjureup/ui/views/deploystatus.py]: +1

**Comment by battlemidget** [comment]
First run looks to complete bootstrap process but the screen seems to not move forward into the application list as seen here:

image

With this in the log:

```
2017-04-17 19:34:13,899 [DEBUG] conjure-up/_unspecified_spell - app.py:216 - LXD version: 2.12, Juju version: 2.2-beta3-xenial-amd64
2017-04-17 19:34:14,498 [INFO] conjure-up/_unspecified_spell - events.py:89 - Watching for shutdown
2017-04-17 19:34:18,934 [DEBUG] conjure-up/kubernetes-core - download.py:60 - Path is local filesystem, copying /home/ubuntu/.cache/conjure-up-spells/kubernetes-core to /home/ubuntu/.cache/conjure-up/kubernetes-core
2017-04-17 19:34:18,948 [DEBUG] conjure-up/kubernetes-core - utils.py:388 - Pulling bundle for kubernetes-core from channel: stable
2017-04-17 19:34:24,922 [DEBUG] conjure-up/kubernetes-core - common.py:124 - Found an IPv4 address, assuming LXD is configured.
2017-04-17 19:34:24,948 [INFO] conjure-up/kubernetes-core - gui.py:73 - Ignoring placement spec because we are deploying to LXD: ['lxd:0']
2017-04-17 19:34:24,948 [INFO] conjure-up/kubernetes-core - gui.py:73 - Ignoring placement spec because we are deploying to LXD: ['0']
2017-04-17 19:34:24,948 [INFO] conjure-up/kubernetes-core - gui.py:73 - Ignoring placement spec because we are deploying to LXD: ['0']
2017-04-17 19:34:24,948 [INFO] conjure-up/kubernetes-core - gui.py:73 - Ignoring placement spec because we are deploying to LXD: ['1']
2017-04-17 19:34:25,056 [DEBUG] conjure-up/kubernetes-core - juju.py:179 - bootstrap cmd: ['juju', 'bootstrap', 'localhost', 'conjure-up-localhost-802', '--default-model', 'conjure-kubernetes-core-25f', '--config', 'image-stream=daily', '--config', 'enable-os-upgrade=false', '--bootstrap-series', 'xenial']
2017-04-17 19:34:25,059 [INFO] conjure-up/kubernetes-core - common.py:22 - Waiting for model connection
2017-04-17 19:34:26,330 [DEBUG] conjure-up/kubernetes-core - gui.py:13 - Rendering bootstrap wait
2017-04-17 19:34:26,440 [DEBUG] conjure-up/kubernetes-core - juju.py:484 - Waiting for pre-deploy
2017-04-17 19:34:26,553 [DEBUG] conjure-up/kubernetes-core - juju.py:484 - Waiting for pre-deploy
2017-04-17 19:34:26,679 [DEBUG] conjure-up/kubernetes-core - juju.py:484 - Waiting for pre-deploy
2017-04-17 19:34:26,794 [DEBUG] conjure-up/kubernetes-core - juju.py:484 - Waiting for pre-deploy
2017-04-17 19:34:26,901 [DEBUG] conjure-up/kubernetes-core - juju.py:484 - Waiting for pre-deploy
2017-04-17 19:36:49,769 [INFO] conjure-up/kubernetes-core - juju.py:125 - Waiting for model conjure-up-localhost-802:conjure-kubernetes-core-25f...
2017-04-17 19:38:45,992 [INFO] conjure-up/kubernetes-core - events.py:95 - Shutting down
```

The may actually be an issue with trying to update the lxd profile?

**Comment by battlemidget** [comment]
Another error when attempting to connect to an existing model:

```
2017-04-17 19:41:54,188 [DEBUG] conjure-up/_unspecified_spell - app.py:216 - LXD version: 2.12, Juju version: 2.2-beta3-xenial-amd64
2017-04-17 19:41:54,712 [INFO] conjure-up/_unspecified_spell - events.py:89 - Watching for shutdown
2017-04-17 19:41:56,936 [DEBUG] conjure-up/kubernetes-core - download.py:60 - Path is local filesystem, copying /home/ubuntu/.cache/conjure-up-spells/kubernetes-core to /home/ubuntu/.cache/conjure-up/kubernetes-core
2017-04-17 19:41:56,945 [DEBUG] conjure-up/kubernetes-core - utils.py:388 - Pulling bundle for kubernetes-core from channel: stable
2017-04-17 19:41:59,815 [INFO] conjure-up/kubernetes-core - gui.py:73 - Ignoring placement spec because we are deploying to LXD: ['lxd:0']
2017-04-17 19:41:59,815 [INFO] conjure-up/kubernetes-core - gui.py:73 - Ignoring placement spec because we are deploying to LXD: ['0']
2017-04-17 19:41:59,815 [INFO] conjure-up/kubernetes-core - gui.py:73 - Ignoring placement spec because we are deploying to LXD: ['0']
2017-04-17 19:41:59,815 [INFO] conjure-up/kubernetes-core - gui.py:73 - Ignoring placement spec because we are deploying to LXD: ['1']
2017-04-17 19:41:59,936 [INFO] conjure-up/kubernetes-core - common.py:22 - Waiting for model connection
2017-04-17 19:42:01,144 [INFO] conjure-up/kubernetes-core - juju.py:125 - Waiting for model conjure-up-localhost-802:conjure-kubernetes-core-fa3...
2017-04-17 19:42:01,144 [INFO] conjure-up/kubernetes-core - juju.py:127 - Connecting to model conjure-up-localhost-802:conjure-kubernetes-core-fa3...
2017-04-17 19:42:01,153 [ERROR] conjure-up/kubernetes-core - events.py:78 - Unhandled exception
Traceback (most recent call last):
  File "/home/ubuntu/conjure-up/conjureup/events.py", line 76, in handle_exception
    raise exc
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/home/ubuntu/conjure-up/conjureup/juju.py", line 730, in add_model
    await login()
  File "/home/ubuntu/conjure-up/conjureup/juju.py", line 128, in login
    await app.juju.client.connect_model(model_name)
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/model.py", line 405, in connect_model
    self.connection = await connection.Connection.connect_model(model_name)
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/client/connection.py", line 270, in connect_model
    model_uuid = models['models'][model_name]['uuid']
KeyError: 'conjure-kubernetes-core-fa3'
2017-04-17 19:42:01,158 [DEBUG] conjure-up/kubernetes-core - __init__.py:27 - Showing dialog for exception: 'conjure-kubernetes-core-fa3'
2017-04-17 19:42:04,280 [INFO] conjure-up/kubernetes-core - events.py:95 - Shutting down
```

**Comment by johnsca** [comment]
@battlemidget I fixed the issue blocking the post-bootstrap, but I'm hitting the `KeyError` when using a model created with `--default-model` on bootstrap.  It seems those don't get added to `models.yaml` which libjuju is depending on to translate the model name to a UUID.  It should be able to query this from the controller, though that may require an extra websocket connection.  I'll look into whether I can refactor that bit better.

**Comment by johnsca** [comment]
I guess I'll also sort out the tests that don't initialize a proper `app` mock.

**Comment by battlemidget** [comment]
@johnsca  with the default model issue, does the same apply if we leave the default model alone and do add model like we discussed for the ability to have suggested model names in the spell? I can work on getting that integrated if that'll make that problem moot.

**Comment by battlemidget** [comment]
@johnsca is this the error you were referring to?

```
2017-04-18 13:00:11,105 [INFO] conjure-up/kubernetes-core - common.py:22 - Waiting for model connection
2017-04-18 13:00:11,122 [DEBUG] conjure-up/kubernetes-core - juju.py:189 - waiting for proc
2017-04-18 13:00:12,953 [DEBUG] conjure-up/kubernetes-core - gui.py:13 - Rendering bootstrap wait
2017-04-18 13:00:13,107 [DEBUG] conjure-up/kubernetes-core - juju.py:488 - Waiting for pre-deploy
2017-04-18 13:00:13,203 [DEBUG] conjure-up/kubernetes-core - juju.py:488 - Waiting for pre-deploy
2017-04-18 13:00:13,304 [DEBUG] conjure-up/kubernetes-core - juju.py:488 - Waiting for pre-deploy
2017-04-18 13:00:13,401 [DEBUG] conjure-up/kubernetes-core - juju.py:488 - Waiting for pre-deploy
2017-04-18 13:00:13,496 [DEBUG] conjure-up/kubernetes-core - juju.py:488 - Waiting for pre-deploy
2017-04-18 13:02:09,659 [DEBUG] conjure-up/kubernetes-core - juju.py:191 - proc done
2017-04-18 13:02:09,660 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Bootstrapped
2017-04-18 13:02:09,660 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting ModelAvailable
2017-04-18 13:02:09,660 [INFO] conjure-up/kubernetes-core - common.py:207 - Bootstrap complete.
2017-04-18 13:02:09,661 [INFO] conjure-up/kubernetes-core - juju.py:128 - Connecting to model conjure-up-localhost-b7c:conjure-kubernetes-core-63b...
2017-04-18 13:02:09,678 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Error
2017-04-18 13:02:09,678 [ERROR] conjure-up/kubernetes-core - events.py:95 - Unhandled exception
Traceback (most recent call last):
  File "/home/ubuntu/conjure-up/conjureup/events.py", line 93, in handle_exception
    raise exc
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/home/ubuntu/conjure-up/conjureup/controllers/newcloud/common.py", line 211, in do_bootstrap
    await juju.login()  # login to the newly created (default) model
  File "/home/ubuntu/conjure-up/conjureup/juju.py", line 129, in login
    await app.juju.client.connect_model(model_name)
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/model.py", line 405, in connect_model
    self.connection = await connection.Connection.connect_model(model_name)
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/client/connection.py", line 270, in connect_model
    model_uuid = models['models'][model_name]['uuid']
KeyError: 'conjure-kubernetes-core-63b'
2017-04-18 13:02:09,682 [DEBUG] conjure-up/kubernetes-core - __init__.py:27 - Showing dialog for exception: 'conjure-kubernetes-core-63b'
2017-04-18 13:04:22,438 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Shutdown
2017-04-18 13:04:22,439 [INFO] conjure-up/kubernetes-core - events.py:112 - Shutting down
```

This happens on a fresh bootstrap and not re-using any controllers

**Comment by battlemidget** [comment]
sla key error with latest commit:
```
2017-04-18 18:56:06,636 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Error
2017-04-18 18:56:06,636 [ERROR] conjure-up/kubernetes-core - events.py:95 - Unhandled exception
Traceback (most recent call last):
  File "/home/ubuntu/conjure-up/conjureup/events.py", line 93, in handle_exception
    raise exc
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/home/ubuntu/conjure-up/conjureup/controllers/newcloud/common.py", line 211, in do_bootstrap
    await juju.login()  # login to the newly created (default) model
  File "/home/ubuntu/conjure-up/conjureup/juju.py", line 129, in login
    await app.juju.client.connect_model(model_name)
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/model.py", line 421, in connect_model
    await self._after_connect()
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/model.py", line 429, in _after_connect
    await self.get_info()
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/model.py", line 572, in get_info
    self.info = await facade.ModelInfo()
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/client/facade.py", line 337, in wrapper
    result = cls.from_json(reply['response'])
  File "/home/ubuntu/conjure-up/conjure-dev/lib/python3.5/site-packages/juju/client/facade.py", line 459, in from_json
    return cls(**d)
TypeError: __init__() got an unexpected keyword argument 'sla'
2017-04-18 18:56:06,643 [DEBUG] conjure-up/kubernetes-core - __init__.py:27 - Showing dialog for exception: __init__() got an unexpected keyword argument 'sla'
```

**Comment by battlemidget** [comment]
nitpick:

image

Can we make the statusbar say like "Relations set, waiting for applications to start"

**Comment by battlemidget** [comment]
All the way to the steps screen with this error:

```
2017-04-18 20:00:19,379 [DEBUG] conjure-up/kubernetes-core - gui.py:120 - Queueing step: <StepWidget: Kubernetes Cluster Status Check>
2017-04-18 20:00:19,380 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Error
2017-04-18 20:00:19,380 [ERROR] conjure-up/kubernetes-core - events.py:95 - Unhandled exception
Traceback (most recent call last):
  File "/home/ubuntu/conjure-up/conjureup/events.py", line 93, in handle_exception
    raise exc
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/home/ubuntu/conjure-up/conjureup/controllers/deploystatus/gui.py", line 14, in _wait_for_applications
    controllers.use('steps').render()
  File "/home/ubuntu/conjure-up/conjureup/controllers/steps/gui.py", line 131, in render
    step_widgets[0].generate_additional_input()
  File "/home/ubuntu/conjure-up/conjureup/ui/widgets/step.py", line 177, in generate_additional_input
    can_sudo = utils.can_sudo()
TypeError: can_sudo() missing 1 required positional argument: 'self'
2017-04-18 20:00:19,382 [DEBUG] conjure-up/kubernetes-core - __init__.py:27 - Showing dialog for exception: can_sudo() missing 1 required positional argument: 'self'
```

**Comment by battlemidget** [comment]
Steps complete now, another small nitpick

image


The statusbar shows the dictionary of the result, maybe we can just say "step is finished" something along those lines

**Comment by battlemidget** [comment]
This test was selecting MAAS for a kubernetes-core deploy:

```
2017-04-18 21:42:48,709 [DEBUG] conjure-up/kubernetes-core - utils.py:404 - Pulling bundle for kubernetes-core from channel: stable
2017-04-18 21:43:23,099 [INFO] conjure-up/kubernetes-core - common.py:183 - Bootstrapping Juju controller.
2017-04-18 21:43:23,100 [DEBUG] conjure-up/kubernetes-core - juju.py:180 - bootstrap cmd: ['juju', 'bootstrap', 'maas/172.16.0.1', 'conjure-up-maas-355', '--default-model', 'conjure-kubernetes-core-3e3', '--config', 'image-stream=daily', '--config', 'enable-os-upgrade=false', '--bootstrap-series', 'xenial', '--credential', 'conjure-up-maas-355']
2017-04-18 21:43:23,105 [INFO] conjure-up/kubernetes-core - common.py:22 - Waiting for model connection
2017-04-18 21:43:23,106 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Error
2017-04-18 21:43:23,106 [ERROR] conjure-up/kubernetes-core - events.py:96 - Unhandled exception
Traceback (most recent call last):
  File "/home/ubuntu/conjure-up/conjureup/events.py", line 94, in handle_exception
    raise exc
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/home/ubuntu/conjure-up/conjureup/controllers/deploy/gui.py", line 274, in connect_maas
    await app.loop.run_in_executor(setup_maas)
TypeError: run_in_executor() missing 1 required positional argument: 'func'
2017-04-18 21:43:23,108 [DEBUG] conjure-up/kubernetes-core - __init__.py:29 - Showing dialog for exception: run_in_executor() missing 1 required positional argument: 'func'
2017-04-18 21:43:23,115 [DEBUG] conjure-up/kubernetes-core - juju.py:189 - waiting for proc
```
After pressing add credential the error showed up

**Comment by battlemidget** [comment]
This next error comes from attempting to pin a service to a machine in the architect view:

```
2017-04-18 23:25:34,815 [DEBUG] conjure-up/kubernetes-core - events.py:18 - Setting Error
2017-04-18 23:25:34,815 [ERROR] conjure-up/kubernetes-core - events.py:96 - Unhandled exception
Traceback (most recent call last):
  File "/home/ubuntu/conjure-up/conjureup/events.py", line 94, in handle_exception
    raise exc
  File "/usr/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "/home/ubuntu/conjure-up/conjureup/controllers/deploy/gui.py", line 236, in _do_deploy
    added_machines = await self.ensure_machines(application)
  File "/home/ubuntu/conjure-up/conjureup/controllers/deploy/gui.py", line 222, in ensure_machines
    await self.get_maas_constraints(virt_machine_id)
  File "/home/ubuntu/conjure-up/conjureup/controllers/deploy/gui.py", line 198, in get_maas_constraints
    None, app.maas.client.assign_id_tags([maas_machine]))
  File "/usr/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
TypeError: 'NoneType' object is not callable
2017-04-18 23:25:34,823 [DEBUG] conjure-up/kubernetes-core - __init__.py:29 - Showing dialog for exception: 'NoneType' object is not callable
```

**Pull request merged**